docs/design.md

```diff
@@ -0,0 +1,156 @@
+cluster but are not guaranteed to work together. That is left as an exercise for
+the chart authors.
+
+Hypper will make it easier to have a repository of charts that work together but
+can be installed separately.
+
+The additional metadata will be stored as Helm chart annotations.
+
+### OCI Storage
+
+OCI artifacts are a new way to store things. Helm supports storing Helm charts
+in OCI distributions as an experimental feature. Many container registries
+support storing non-images in them. The list includes Azure, AWS, Google Cloud,
+GitHub Container Registry, and Harbor. The one notable exception is Docker Hub.
+
+Kubernetes already needs to have a container registry. In air gapped, on-premise,
+and custom environments you need a registry to store your images. If that can be
+used to store charts instead of a separate system you have fewer system
+dependencies.
+
+Hypper will make OCI registries a primary element in the system.
+
+One notable disadvantage to using an OCI registry for storing charts is a lack of
+an `index.yaml` file. This file has metadata enabling some functionality that
+can't be duplicated with an OCI registry due to APIs not existing (e.g., search).
+To restore these features we are looking to put an index into an OCI registry.
+
+### Name As Chart Identifier
+
+Zypper, the package manager used for openSUSE and other Linux distributions,
+uses the package name as the identifier for it. You can change from one vendor
+to another for a package with some differences but still the same package.
+
+Helm does not make this form of assumption. It could be true. For example, you
+could install a chart from one location and then update from a fork in another
+location. Or two charts with the same name could be completely different. It
+is the responsibility of the end user to make this decision.
+
+Hypper is going to make the assumption that charts of the same name are the
+same chart from different locations.
+
+### Performance
+
+When working with many files over a network there are opportunities to use
+caching, features of protocols, and design choices to improve performance. The
+following examples illustrate this:
+
+- JSON is faster to parse and uses less memory than YAML does. An `index.json`
+  file would faster and less memory intensive for Helm to work with than
+  `index.yaml`. A change like this could be done in a backwards compatible
+  manner.
+- `index.yaml` files are regularly pulled down to get updates and this is a
+  manual step. Using HTTP headers we can check if the content changed and only
+  pull it in those cases. By doing that some of the updates can be automated,
+  as well. A header that could be used, for example, is `etag`.
+- Charts can be stored locally in a cache and used from there rather than being
+  downloaded each time. This cache could be stored in places like a CI system
+  (e.g., like the way packages are stored for programming languages) or on a
+  local development computer.
```

Comment:
What do you mean by this? Do you mean that hypper needs to understand Cache expiry/cache invalidation of OCI/webservers? Or do you mean that hypper needs to implement a "download (into cache)" function for air-gapped?

Reply:
Hypper can have a local cache directory where it stores charts. If you already have the chart at a version in the cache than you can install it right from there rather than downloading it. The index has a digest (sha256) so we can likely make this a content addressable system. This could be something to make work better in upstream Helm.